base/strings/util.jl

```diff
@@ -633,30 +633,36 @@ julia> hex2bytes(a)
 function hex2bytes end
 
 hex2bytes(s::AbstractString) = hex2bytes(String(s))
-hex2bytes(s::Union{String,AbstractVector{UInt8}}) = hex2bytes!(Vector{UInt8}(undef, length(s) >> 1), s)
-
-_firstbyteidx(s::String) = 1
-_firstbyteidx(s::AbstractVector{UInt8}) = first(eachindex(s))
-_lastbyteidx(s::String) = sizeof(s)
-_lastbyteidx(s::AbstractVector{UInt8}) = lastindex(s)
+hex2bytes(s::String) = hex2bytes(transcode(UInt8, s))
```

Comment:
It is better to do this specialization within `hex2bytes!`, since the user may call `hex2bytes!` directly. 

On the other hand, the handling of non-`String` `AbstractString` is a matter of debate. In terms of generality, it might be better to leave it to the iterator implementation and add the following fallback chain.
```julia
number_from_hex(c::AbstractChar) = number_from_hex(Char(c))
number_from_hex(c::Char) = number_from_hex(UInt8(c))
```
However, in practice, I think the majority of non-`String` `AbstractString`s is `SubString{String}`.

Reply:
I wasn't sure whether that was needed at all, but without the `transcode`, I immediately run into a regression across the board of + 20±6% on my machine when calling `hex2bytes(::String)`. Presumably that's because of the necessary checks from hitting `UInt8(::Char)`, since not all of them are convertible to UInt8. I originally decided to put them in since `hex2bytes` is only expected to work on correct bytes anyway and since `number_from_hex` checks them already.

If a user decides to call `hex2bytes!` with a string directly, they'll hit `nfh(::Char)` and hit the slow conversion that way - I guess I could add another `hex2bytes!(dest, s::String) = hex2bytes!(dest, transcode(UInt8, String))` method, which would allow us to remove `nfh(::Char)`?

Reply:
Moving `transcode` to `hex2bytes!` and removing it from the non-mutating version can lead to subtle allocation differences if Unicode is involved (which would throw anyway), leading to misleading error messages being printed.

E.g., just having

```julia
h2b(s) = h2b!(Vector{UInt8}(undef, length(s) >> 1), s)

h2b!(dest::AbstractArray{UInt8}, s::String) = h2b!(dest, transcode(UInt8, s))
function h2b!(dest::AbstractArray{UInt8}, itr)
    isodd(length(itr)) && throw(ArgumentError("length of iterable must be even"))
    @boundscheck 2*length(dest) < length(itr) && throw(ArgumentError("length of output array must be at least half of the length of input iterable"))
```

and calling `h2b("±±±")` would hit the boundscheck instead of the non-ASCII check in `nfh`. This could be solved by using `sizeof` for strings, but that disallows other iterables (the whole point of this PR).

Reply:
If you want to improve the error message, it might look like this, for example:
```julia
function h2b!(dest::AbstractArray{UInt8}, s::String)
    if sizeof(s) != length(s)
        throw(ArgumentError("input string must consist of hexadecimal characters only."))
    end
    h2b!(dest, transcode(UInt8, s))
end
```

Reply:
If you apply the suggestion above, I think it would be more consistent to remove this line,  as I initially suggested.
```julia
hex2bytes(s::String) = hex2bytes(transcode(UInt8, s))
```
Also, in order to handle `AbstractString` properly, I think you need to add the following, or add a forced conversion to `String`.
```julia
number_from_hex(c::AbstractChar) = number_from_hex(Char(c))
```
(For hygiene reasons, I think it is better to keep the name `number_from_hex` than `nfh`.)

I'm not good at English, so please refer to advices from others regarding the error message. :sweat_smile:

Reply:
Yeah, with the `sizeof` check and passthrough we already take care of the `transcode`, good point.

I think the text is perfectly fine & concise, that's why I added it. I wouldn't have written it any other way :)